external-crates/move/crates/move-compiler/src/sui_mode/linters/custom_rules/redundant_ref_deref.rs

```diff
@@ -0,0 +1,73 @@
+// Implements lint rule for Move code to detect redundant reference dereference patterns.
+// It identifies and reports unnecessary temporary borrow followed by a dereference and a local borrow.
+// Aims to improve code efficiency by suggesting direct usage of expressions without redundant operations.
+use crate::{
+    diag,
+    diagnostics::{
+        codes::{custom, DiagnosticInfo, Severity},
+        WarningFilters,
+    },
+    shared::{program_info::TypingProgramInfo, CompilationEnv},
+    sui_mode::linters::{LinterDiagCategory, LINTER_DEFAULT_DIAG_CODE, LINT_WARNING_PREFIX},
+    typing::{
+        ast::{self as T, UnannotatedExp_},
+        visitor::{TypingVisitorConstructor, TypingVisitorContext},
+    },
+};
+use move_ir_types::location::Loc;
+
+const REDUNDANT_REF_DEREF_DIAG: DiagnosticInfo = custom(
+    LINT_WARNING_PREFIX,
+    Severity::Warning,
+    LinterDiagCategory::RedundantRefDeref as u8,
+    LINTER_DEFAULT_DIAG_CODE,
+    "",
+);
+
+pub struct RedundantRefDerefVisitor;
+
+pub struct Context<'a> {
+    env: &'a mut CompilationEnv,
+}
+
+impl TypingVisitorConstructor for RedundantRefDerefVisitor {
+    type Context<'a> = Context<'a>;
+
+    fn context<'a>(
+        env: &'a mut CompilationEnv,
+        _program_info: &'a TypingProgramInfo,
+        _program: &T::Program_,
+    ) -> Self::Context<'a> {
+        Context { env }
+    }
+}
+
+impl TypingVisitorContext for Context<'_> {
+    fn visit_exp_custom(&mut self, exp: &mut T::Exp) -> bool {
+        // If the expression is a temporary borrow followed by a dereference that is itself a local borrow,
+        // it indicates a redundant ref-deref pattern.
+        if let UnannotatedExp_::TempBorrow(_, borrow_exp) = &exp.exp.value {
+            if let UnannotatedExp_::Dereference(deref_exp) = &borrow_exp.exp.value {
+                if let UnannotatedExp_::BorrowLocal(_, _) = &deref_exp.exp.value {
```

Comment:
You can use let-else to simplify the nesting here

Reply:
refactored